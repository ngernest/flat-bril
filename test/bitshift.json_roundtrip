{
  "functions": [
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "n"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v2",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v1",
            "v2"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v3"
          ],
          "funcs": [],
          "labels": [
            "then.0",
            "else.0"
          ],
          "op": "br"
        },
        {
          "label": "then.0"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v4"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "else.0"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v5",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "n"
          ],
          "dest": "v6",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v7",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "v6",
            "v7"
          ],
          "dest": "v8",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "v5",
            "v8"
          ],
          "dest": "half",
          "funcs": [
            "pow"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "half"
          ],
          "dest": "half",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "half"
          ],
          "dest": "v9",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "half"
          ],
          "dest": "v10",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v9",
            "v10"
          ],
          "dest": "v11",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v11"
          ],
          "dest": "half2",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "n"
          ],
          "dest": "v13",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v14",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "v13",
            "v14"
          ],
          "dest": "v15",
          "funcs": [
            "mod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "dest": "v16",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v15",
            "v16"
          ],
          "dest": "v17",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v17"
          ],
          "funcs": [],
          "labels": [
            "then.12",
            "else.12"
          ],
          "op": "br"
        },
        {
          "label": "then.12"
        },
        {
          "args": [
            "half2"
          ],
          "dest": "v18",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v19",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v18",
            "v19"
          ],
          "dest": "v20",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v20"
          ],
          "dest": "ans",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "endif.12"
          ],
          "op": "jmp"
        },
        {
          "label": "else.12"
        },
        {
          "args": [
            "half2"
          ],
          "dest": "v21",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v21"
          ],
          "dest": "ans",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "endif.12"
        },
        {
          "args": [
            "ans"
          ],
          "dest": "v22",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v22"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "pow",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "a"
          ],
          "dest": "v0",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "a"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "b"
          ],
          "dest": "v2",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v1",
            "v2"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "b"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v3",
            "v4"
          ],
          "dest": "v5",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v0",
            "v5"
          ],
          "dest": "v6",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v6"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "mod",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "step",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "v0",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "step"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v0",
            "v1"
          ],
          "dest": "p",
          "funcs": [
            "pow"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "p"
          ],
          "dest": "p",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v2",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "p"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v2",
            "v3"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v4"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "LEFTSHIFT",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "step",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "v0",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "step"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v0",
            "v1"
          ],
          "dest": "p",
          "funcs": [
            "pow"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "p"
          ],
          "dest": "p",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v2",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "p"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v2",
            "v3"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "v4"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "RIGHTSHIFT",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        },
        {
          "name": "c",
          "type": "int"
        },
        {
          "name": "d",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "a"
          ],
          "dest": "v2",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "b"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v2",
            "v3"
          ],
          "dest": "ans1",
          "funcs": [
            "LEFTSHIFT"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "ans1"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "args": [
            "c"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "d"
          ],
          "dest": "v5",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v4",
            "v5"
          ],
          "dest": "ans2",
          "funcs": [
            "RIGHTSHIFT"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "ans2"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        }
      ],
      "name": "main"
    }
  ]
}
