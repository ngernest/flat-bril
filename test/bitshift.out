@pow
	op: id   		dest: v1		type: int		args: ["n"]	
	op: const		dest: v2		type: int		value: 1	
	op: eq   		dest: v3		type: bool		args: ["v1", "v2"]	
	op: br   		args: ["v3"]		labels: ["then.0", "else.0"]
.then.0
	op: id   		dest: v4		type: int		args: ["x"]	
	op: ret  		args: ["v4"]	
.else.0
	op: id   		dest: v5		type: int		args: ["x"]	
	op: id   		dest: v6		type: int		args: ["n"]	
	op: const		dest: v7		type: int		value: 2	
	op: div  		dest: v8		type: int		args: ["v6", "v7"]	
	op: call 		dest: half		type: int		args: ["v5", "v8"]		funcs: ["pow"]
	op: id   		dest: half		type: int		args: ["half"]	
	op: id   		dest: v9		type: int		args: ["half"]	
	op: id   		dest: v10		type: int		args: ["half"]	
	op: mul  		dest: v11		type: int		args: ["v9", "v10"]	
	op: id   		dest: half2		type: int		args: ["v11"]	
	op: id   		dest: v13		type: int		args: ["n"]	
	op: const		dest: v14		type: int		value: 2	
	op: call 		dest: v15		type: int		args: ["v13", "v14"]		funcs: ["mod"]
	op: const		dest: v16		type: int		value: 1	
	op: eq   		dest: v17		type: bool		args: ["v15", "v16"]	
	op: br   		args: ["v17"]		labels: ["then.12", "else.12"]
.then.12
	op: id   		dest: v18		type: int		args: ["half2"]	
	op: id   		dest: v19		type: int		args: ["x"]	
	op: mul  		dest: v20		type: int		args: ["v18", "v19"]	
	op: id   		dest: ans		type: int		args: ["v20"]	
	op: jmp  		labels: ["endif.12"]
.else.12
	op: id   		dest: v21		type: int		args: ["half2"]	
	op: id   		dest: ans		type: int		args: ["v21"]	
.endif.12
	op: id   		dest: v22		type: int		args: ["ans"]	
	op: ret  		args: ["v22"]	
@mod
	op: id   		dest: v0		type: int		args: ["a"]	
	op: id   		dest: v1		type: int		args: ["a"]	
	op: id   		dest: v2		type: int		args: ["b"]	
	op: div  		dest: v3		type: int		args: ["v1", "v2"]	
	op: id   		dest: v4		type: int		args: ["b"]	
	op: mul  		dest: v5		type: int		args: ["v3", "v4"]	
	op: sub  		dest: v6		type: int		args: ["v0", "v5"]	
	op: ret  		args: ["v6"]	
@LEFTSHIFT
	op: const		dest: v0		type: int		value: 2	
	op: id   		dest: v1		type: int		args: ["step"]	
	op: call 		dest: p 		type: int		args: ["v0", "v1"]		funcs: ["pow"]
	op: id   		dest: p 		type: int		args: ["p"]	
	op: id   		dest: v2		type: int		args: ["x"]	
	op: id   		dest: v3		type: int		args: ["p"]	
	op: mul  		dest: v4		type: int		args: ["v2", "v3"]	
	op: ret  		args: ["v4"]	
@RIGHTSHIFT
	op: const		dest: v0		type: int		value: 2	
	op: id   		dest: v1		type: int		args: ["step"]	
	op: call 		dest: p 		type: int		args: ["v0", "v1"]		funcs: ["pow"]
	op: id   		dest: p 		type: int		args: ["p"]	
	op: id   		dest: v2		type: int		args: ["x"]	
	op: id   		dest: v3		type: int		args: ["p"]	
	op: div  		dest: v4		type: int		args: ["v2", "v3"]	
	op: ret  		args: ["v4"]	
@main
	op: id   		dest: v2		type: int		args: ["a"]	
	op: id   		dest: v3		type: int		args: ["b"]	
	op: call 		dest: ans1		type: int		args: ["v2", "v3"]		funcs: ["LEFTSHIFT"]
	op: print		args: ["ans1"]	
	op: id   		dest: v4		type: int		args: ["c"]	
	op: id   		dest: v5		type: int		args: ["d"]	
	op: call 		dest: ans2		type: int		args: ["v4", "v5"]		funcs: ["RIGHTSHIFT"]
	op: print		args: ["ans2"]	
